pkg/proxy/accesslog/log.go

```
@@ -98,38 +149,90 @@ func Log(l *LogRecord, typ FlowType, verdict FlowVerdict, code int) {
 	l.Verdict = verdict
 	l.Metadata = metadata
 
-	l.HTTP = &LogRecordHTTP{
-		Code:     code,
-		Method:   l.Request.Method,
-		URL:      l.Request.URL,
-		Protocol: l.Request.Proto,
-		Headers:  l.Request.Header,
-	}
+	switch L7type {
+	case L7TypeHTTP:
+		{
+			l.HTTP = &LogRecordHTTP{
+				Code:     code,
+				Method:   l.HttpRequest.Method,
+				URL:      l.HttpRequest.URL,
+				Protocol: l.HttpRequest.Proto,
+				Headers:  l.HttpRequest.Header,
+			}
+			log.WithFields(log.Fields{
+				FieldType:     typ,
+				FieldVerdict:  verdict,
+				FieldCode:     code,
+				FieldMethod:   l.HttpRequest.Method,
+				FieldURL:      l.HttpRequest.URL,
+				FieldProtocol: l.HttpRequest.Proto,
+				FieldHeader:   l.HttpRequest.Header,
+				FieldFilePath: logPath,
+			}).Debug("Logging HTTP L7 flow record")
+
+			if logBuf == nil {
+				log.WithFields(log.Fields{
+					FieldFilePath: logPath,
+				}).Debug("Skipping writing to access log (write buffer nil)")
+				return
+			}
+
+			b, err := json.Marshal(*l)
+			if err != nil {
+				logString(err.Error(), true)
+			} else {
+				logString(string(b), true)
+			}
+		}
+	case L7TypeKafka:
+		{
+			l.Kafka = &LogRecordKafka{
+				Code:          code,
+				APIVersion:    l.KafkaRequest.GetVersion(),
+				APIKey:        APIKeyMap[l.KafkaRequest.GetAPIKey()],
```

This should be done via lookup function that check out of bounds access to avoid out of bounds slice access panics.

---

Worse, the map lookup will return the nil string value (so empty string) so this would be a silent error.

---

will do